File: employee-service-ruby/models/employee.rb
require 'securerandom'
require_relative '../config/logger'

# Employee model class for managing employee data
class Employee
  attr_reader :id, :first_name, :last_name, :email, :department, :salary

  def initialize(id, first_name, last_name, email, department, salary)
    @id = id || SecureRandom.uuid
    @first_name = first_name
    @last_name = last_name
    @email = email
    @department = department
    @salary = salary
  end

  # Convert employee object to hash
  def to_hash
    {
      id: @id,
      firstName: @first_name,
      lastName: @last_name,
      email: @email,
      department: @department,
      salary: @salary
    }
  end
end

# Employee repository for CRUD operations
class EmployeeRepository
  def initialize
    # Initialize in-memory database with sample data
    @employees = [
      Employee.new('1', 'John', 'Doe', 'john.doe@example.com', 'IT', 75000.0),
      Employee.new('2', 'Jane', 'Smith', 'jane.smith@example.com', 'HR', 65000.0),
      Employee.new('3', 'Michael', 'Brown', 'michael.brown@example.com', 'Finance', 85000.0),
      Employee.new('4', 'Sarah', 'Johnson', 'sarah.johnson@example.com', 'Marketing', 70000.0),
      Employee.new('5', 'David', 'Williams', 'david.williams@example.com', 'Operations', 72000.0)
    ]
    
    Logger.info('Initialized in-memory employee database with sample data')
  end

  # Get all employees
  def find_all
    Logger.debug('Retrieving all employees from in-memory database')
    @employees.map(&:to_hash)
  end

  # Find employee by ID
  def find_by_id(id)
    Logger.debug("Finding employee with ID: #{id}")
    employee = @employees.find { |emp| emp.id == id }
    employee&.to_hash
  end

  # Create a new employee
  def create(data)
    employee = Employee.new(
      nil,
      data['firstName'],
      data['lastName'],
      data['email'],
      data['department'],
      data['salary']
    )
    
    @employees << employee
    Logger.info("Created new employee with ID: #{employee.id}")
    
    employee.to_hash
  end

  # Update an existing employee
  def update(id, data)
    Logger.debug("Attempting to update employee with ID: #{id}")
    
    index = @employees.find_index { |emp| emp.id == id }
    return nil if index.nil?
    
    current = @employees[index]
    employee = Employee.new(
      id,
      data['firstName'] || current.first_name,
      data['lastName'] || current.last_name,
      data['email'] || current.email,
      data['department'] || current.department,
      data['salary'] || current.salary
    )
    
    @employees[index] = employee
    Logger.info("Successfully updated employee with ID: #{id}")
    
    employee.to_hash
  end

  # Delete an employee
  def delete(id)
    Logger.debug("Attempting to delete employee with ID: #{id}")
    
    initial_size = @employees.size
    @employees.reject! { |emp| emp.id == id }
    
    deleted = @employees.size < initial_size
    
    if deleted
      Logger.info("Successfully deleted employee with ID: #{id}")
    else
      Logger.warn("Employee with ID: #{id} not found for deletion")
    end
    
    deleted
  end
end

File: employee-service-go/models/employee.go
package models

import (
	"employee-service/utils"
	"github.com/google/uuid"
	"sync"
)

// Employee represents an employee entity
type Employee struct {
	ID         string  `json:"id"`
	FirstName  string  `json:"firstName" binding:"required"`
	LastName   string  `json:"lastName" binding:"required"`
	Email      string  `json:"email" binding:"required,email"`
	Department string  `json:"department"`
	Salary     float64 `json:"salary"`
}

// EmployeeRepository manages the employee data
type EmployeeRepository struct {
	employees map[string]Employee
	mutex     sync.RWMutex
}

// Global instance of the repository
var (
	employeeRepo *EmployeeRepository
	once         sync.Once
)

// InitEmployeeRepository initializes the employee repository with sample data
func InitEmployeeRepository() *EmployeeRepository {
	once.Do(func() {
		employeeRepo = &EmployeeRepository{
			employees: make(map[string]Employee),
			mutex:     sync.RWMutex{},
		}

		// Add sample data
		sampleEmployees := []Employee{
			{
				ID:         "1",
				FirstName:  "John",
				LastName:   "Doe",
				Email:      "john.doe@example.com",
				Department: "IT",
				Salary:     75000.0,
			},
			{
				ID:         "2",
				FirstName:  "Jane",
				LastName:   "Smith",
				Email:      "jane.smith@example.com",
				Department: "HR",
				Salary:     65000.0,
			},
			{
				ID:         "3",
				FirstName:  "Michael",
				LastName:   "Brown",
				Email:      "michael.brown@example.com",
				Department: "Finance",
				Salary:     85000.0,
			},
			{
				ID:         "4",
				FirstName:  "Sarah",
				LastName:   "Johnson",
				Email:      "sarah.johnson@example.com",
				Department: "Marketing",
				Salary:     70000.0,
			},
			{
				ID:         "5",
				FirstName:  "David",
				LastName:   "Williams",
				Email:      "david.williams@example.com",
				Department: "Operations",
				Salary:     72000.0,
			},
		}

		for _, emp := range sampleEmployees {
			employeeRepo.employees[emp.ID] = emp
		}

		utils.Log.Info("Initialized in-memory employee database with sample data")
	})

	return employeeRepo
}

// GetEmployeeRepository returns the singleton instance of the employee repository
func GetEmployeeRepository() *EmployeeRepository {
	if employeeRepo == nil {
		InitEmployeeRepository()
	}
	return employeeRepo
}

// GetAll returns all employees
func (r *EmployeeRepository) GetAll() []Employee {
	r.mutex.RLock()
	defer r.mutex.RUnlock()

	utils.Log.Debug("Retrieving all employees")
	
	employees := make([]Employee, 0, len(r.employees))
	for _, emp := range r.employees {
		employees = append(employees, emp)
	}
	
	utils.Log.Infof("Retrieved %d employees successfully", len(employees))
	return employees
}

// GetByID returns an employee by ID
func (r *EmployeeRepository) GetByID(id string) (Employee, bool) {
	r.mutex.RLock()
	defer r.mutex.RUnlock()

	utils.Log.Debugf("Looking for employee with ID: %s", id)
	
	employee, exists := r.employees[id]
	
	if !exists {
		utils.Log.Warnf("Employee with ID %s not found", id)
	} else {
		utils.Log.Infof("Found employee with ID: %s", id)
	}
	
	return employee, exists
}

// Create adds a new employee
func (r *EmployeeRepository) Create(employee Employee) Employee {
	r.mutex.Lock()
	defer r.mutex.Unlock()

	utils.Log.Debug("Creating new employee")
	
	// Generate ID if not provided
	if employee.ID == "" {
		employee.ID = uuid.New().String()
	}
	
	r.employees[employee.ID] = employee
	
	utils.Log.Infof("Created employee with ID: %s", employee.ID)
	return employee
}

// Update modifies an existing employee
func (r *EmployeeRepository) Update(id string, employee Employee) (Employee, bool) {
	r.mutex.Lock()
	defer r.mutex.Unlock()

	utils.Log.Debugf("Attempting to update employee with ID: %s", id)
	
	_, exists := r.employees[id]
	if !exists {
		utils.Log.Warnf("Cannot update: employee with ID %s not found", id)
		return Employee{}, false
	}
	
	// Preserve the ID
	employee.ID = id
	r.employees[id] = employee
	
	utils.Log.Infof("Updated employee with ID: %s", id)
	return employee, true
}

// Delete removes an employee
func (r *EmployeeRepository) Delete(id string) bool {
	r.mutex.Lock()
	defer r.mutex.Unlock()

	utils.Log.Debugf("Attempting to delete employee with ID: %s", id)
	
	_, exists := r.employees[id]
	if !exists {
		utils.Log.Warnf("Cannot delete: employee with ID %s not found", id)
		return false
	}
	
	delete(r.employees, id)
	
	utils.Log.Infof("Deleted employee with ID: %s", id)
	return true
}
